frontend/readline.c

```diff
@@ -36,14 +42,53 @@ static void sigint_handler(int n) {
 }
 #endif
 
+static int mkdir_all(const char *path) {
+	struct stat st = {0};
+	if (stat(path, &st) == 0 && S_ISDIR(st.st_mode)) {
+		return 0;
+	}
+
+	char *subpath = strdup(path);
+	subpath = dirname(subpath);
+
+	if (strlen(subpath) > 1) {
+		int ret_val = mkdir_all(subpath);
+		if (ret_val != 0) {
+			return ret_val;
+		}
+	}
+
+	int ret_val = mkdir(path, 0700);
+	free(subpath);
+	return ret_val;
+}
+
 static char *get_history_path(void) {
-	const char *home = getenv("HOME");
-	int len = snprintf(NULL, 0, "%s/.mrsh_history", home);
-	char *path = malloc(len + 1);
+	char *xdg_data_home = getenv("XDG_DATA_HOME");
```

Comment:
I suggest to use `XDG_STATE_HOME`.

_The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:_

_actions history (logs, **history**, recently used files, ...)_

Ref: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Reply:
This is indeed a good suggestion. However, it got added to the spec in November 2020 and the new spec was published in May 2021, which this PR predates by quite a bit.

So it makes sense the PR didn't originally use it, but this is a good opportunity to handle it right from the start...